### ISSUE TITLE
0.13: [Bug]: Empty source folder for build dependencies when running `garden deploy`, but not `garden build`

### ISSUE THREAD

**Issue by spenczar** [garden-io/garden#4505]
### Current Behavior

I'm not 100% certain of the cause, but while upgrading Modules to Actions, I have repeatedly encountered error messages indicating an empty build directory. I can confirm that `.garden/sources/actions/<my-action>` is created as an empty directory in these cases.

Before migration, an example particular Module looked like the following. It has a module for building a container, and a second module for deploying a helm chart for the container. It uses a `build.dependencies[]` and references the build module's output:
```yml
kind: Module
type: container
name: gateway-cache-image
description: Image used by gateway-cache
repositoryUrl: git@github.com:B612-Asteroid-Institute/gateway.git#main
dockerfile: ./Dockerfile
services: []
exclude: []
---
kind: Module
type: helm
name: gateway-cache
description: Deploys the simple caching service
repositoryUrl: git@github.com:B612-Asteroid-Institute/gateway.git#main
chartPath: ./chart
exclude: []
build:
  dependencies:
    - gateway-cache-image
values:
  image:
    repository: ${modules.gateway-cache-image.outputs.deployment-image-name}
    tag: ${modules.gateway-cache-image.version}
  hostname: gateway-cache.${var.hostname}
```

I rewrote that to this:
```yml
kind: Build
name: gateway-cache
description: Image used by gateway-cache
type: container
source:
  repository:
    url: git@github.com:B612-Asteroid-Institute/gateway.git#main
---

kind: Deploy
name: gateway-cache
description: Deploys the simple caching service
type: helm
build: gateway-cache
source:
  repository:
    url: git@github.com:B612-Asteroid-Institute/candidate_review.git#main
spec:
  chart:
    path: ./chart
  values:
    image:
      repository: ${actions.build.gateway-cache.outputs.deployment-image-name}
      tag: ${actions.build.gateway-cache.version}
    hostname: gateway-cache.${var.hostname}
```

Then I ran `garden deploy gateway-cache`, and I got the following _build_ error:

```
$ garden deploy gateway-cache
Deploy 🚀

Garden v0.13 (Bonsai) is a major release with significant changes. Please help us improve it by reporting any issues/bugs here:
https://go.garden.io/report-bonsai
→ Run garden util hide-warning 0.13-bonsai to disable this warning.
ℹ garden               → Running in Garden environment dev.ai-swnelson
ℹ providers            → Getting status...
✔ providers            → Cached (took 0.6 sec)
ℹ providers            → Run with --force-refresh to force a refresh of provider statuses.
ℹ graph                → Resolving actions and modules...
✔ graph                → Done (took 1.1 sec)
✔ build.gateway-cache  → Connected to sync target Build sync Pod
ℹ build.gateway-cache  → [sync]: Sync disconnected
ℹ build.gateway-cache  → [sync]: Scanning files to sync
✔ build.gateway-cache  → [sync]: Completed initial sync from Module gateway-cache build path to Build sync Pod
ℹ build.gateway-cache  → [sync]: Watching for changes
ℹ build.gateway-cache  → [sync]: Scanning files to sync
ℹ build.gateway-cache  → [sync]: Synchronized from Module gateway-cache build path to Build sync Pod at 2:02:53 PM
ℹ build.gateway-cache  → [sync]: Scanning files to sync
ℹ build.gateway-cache  → [sync]: Synchronized from Module gateway-cache build path to Build sync Pod at 2:02:53 PM
ℹ build.gateway-cache  → [sync]: Scanning files to sync
ℹ build.gateway-cache  → [sync]: Synchronized from Module gateway-cache build path to Build sync Pod at 2:02:53 PM
ℹ build.gateway-cache  → File sync to cluster complete
ℹ build.gateway-cache  → Buildkit Building image gateway-cache:v-89d107b987...
✖ build.gateway-cache  → Build failed (took 11.5 sec)
✖ build.gateway-cache  → Failed processing Build type=container name=gateway-cache (took 11.53 sec). Here is the output:

────────────────────────────────────────────────────────────────────────────────
Command exited with code 1:
#1 [internal] load .dockerignore
#1 transferring context: 2B done
#1 DONE 0.0s

#2 [internal] load build definition from Dockerfile
#2 transferring dockerfile: 2B done
#2 DONE 0.0s
error: failed to solve: failed to read dockerfile: open /tmp/buildkit-mount350800148/Dockerfile: no such file or directory
────────────────────────────────────────────────────────────────────────────────
1 deploy action(s) failed!
```

This was confusing, but I eventually figured out that the `.garden/sources/action/build.gateway-cache--*` directory was totally empty:

```
ls -alh .garden/sources/action/build.gateway-cache--f30ff7e3aa
total 0
drwxr-xr-x  2 swnelson  staff    64B May 31 14:02 .
drwxr-xr-x  7 swnelson  staff   224B May 31 14:02 ..
```

Then I ran `garden build gateway-cache` to _explicitly_ build it, and it worked!:
```
$ garden build gateway-cache
Build 🔨

Garden v0.13 (Bonsai) is a major release with significant changes. Please help us improve it by reporting any issues/bugs here:
https://go.garden.io/report-bonsai
→ Run garden util hide-warning 0.13-bonsai to disable this warning.
ℹ garden               → Running in Garden environment dev.ai-swnelson
ℹ providers            → Getting status...
✔ providers            → Cached (took 0.7 sec)
ℹ providers            → Run with --force-refresh to force a refresh of provider statuses.
ℹ graph                → Resolving actions and modules...
✔ graph                → Done (took 1.1 sec)
✔ build.gateway-cache  → Connected to sync target Build sync Pod
ℹ build.gateway-cache  → [sync]: Scanning files to sync
ℹ build.gateway-cache  → [sync]: Reconciling sync changes
ℹ build.gateway-cache  → [sync]: Staging files to sync in target
ℹ build.gateway-cache  → [sync]: Syncing changes...
ℹ build.gateway-cache  → [sync]: Saving sync archive
✔ build.gateway-cache  → [sync]: Completed initial sync from Module gateway-cache build path to Build sync Pod
ℹ build.gateway-cache  → [sync]: Watching for changes
ℹ build.gateway-cache  → [sync]: Scanning files to sync
ℹ build.gateway-cache  → [sync]: Reconciling sync changes
ℹ build.gateway-cache  → [sync]: Synchronized from Module gateway-cache build path to Build sync Pod at 2:08:18 PM
ℹ build.gateway-cache  → File sync to cluster complete
ℹ build.gateway-cache  → Buildkit Building image gateway-cache:v-fc955ff417...
✔ build.gateway-cache  → Done (took 27.7 sec)
```

Was the original source folder populated? No, but a sibling with a different hash _was_ populated:
```
$ ls -alh .garden/sources/action/build.gateway-cache--f30ff7e3aa
total 0
drwxr-xr-x  2 swnelson  staff    64B May 31 14:02 .
drwxr-xr-x  8 swnelson  staff   256B May 31 14:07 ..

$ ls -alh .garden/sources/action/build.gateway-cache--390814b07d
total 64
drwxr-xr-x  13 swnelson  staff   416B May 31 14:07 .
drwxr-xr-x   8 swnelson  staff   256B May 31 14:07 ..
drwxr-xr-x  13 swnelson  staff   416B May 31 14:07 .git
-rw-r--r--   1 swnelson  staff   857B May 31 14:07 .pre-commit-config.yaml
-rw-r--r--   1 swnelson  staff   565B May 31 14:07 Dockerfile
-rw-r--r--   1 swnelson  staff   397B May 31 14:07 README.md
drwxr-xr-x   5 swnelson  staff   160B May 31 14:07 chart
drwxr-xr-x   6 swnelson  staff   192B May 31 14:07 gateway
-rw-r--r--   1 swnelson  staff   113B May 31 14:07 pyproject.toml
-rw-r--r--   1 swnelson  staff   1.1K May 31 14:07 requirements.txt
-rw-r--r--   1 swnelson  staff   583B May 31 14:07 setup.cfg
-rw-r--r--   1 swnelson  staff    38B May 31 14:07 setup.py
-rw-r--r--   1 swnelson  staff    86B May 31 14:07 uwsgi.ini
```

### Expected behavior
I expect that when I run `garden deploy`, if it has any explicit build dependencies which depend on remote sources, those remote sources will get populated in exactly the same way as with `garden build`. 

### Reproducible example

I'm not sure I have a minimal reproducer, but the above may help.

### Workaround

Run `garden build` explicitly before every deploy.

### Suggested solution(s)

I don't know

### Additional context

I think this is only happening with

### Your environment

* **OS:** macOS
* **How I'm running Kubernetes:** GKE

$ garden version
0.13.0

**Comment by stefreak** [comment]
Similar issue I discovered recently with actions: #4506 – I strongly feel this might be related.

**Comment by Orzelius** [comment]
Hey @spenczar, thanks for reporting this. It might have been fixed via https://github.com/garden-io/garden/pull/4516, could you update to `0.13.7`and try again please.

**Comment by spenczar** [comment]
@Orzelius Yes, updating to 0.13.7 fixed my issue. Thanks.

**Comment by vvagaytsev** [comment]
Thanks, @spenczar! We're closing this as done via #4516.

**Comment by vvagaytsev** [comment]
This was fixed in `0.13.7` (see #4516), but that PR was partially reverted in `0.13.8`. Reopening this as a valid one.

**Comment by stefreak** [comment]
This bug can be reproduced by re-enabling and running a skipped test `yarn integ-remote -g 'builds a Docker image and emits a namespace status event'`

It will fail with the error 
```
  1) ensureBuildkit
       cluster-buildkit mode
         builds a Docker image and emits a namespace status event:
     Error: Command exited with code 1:
#1 [internal] load .dockerignore
#1 transferring context: 2B 0.0s done
#1 DONE 0.0s

#2 [internal] load build definition from Dockerfile
#2 transferring dockerfile: 2B 0.0s done
#2 DONE 0.0s
error: failed to solve: failed to read dockerfile: open /tmp/buildkit-mount4078873673/Dockerfile: no such file or directory
```

And this is because the mutagen sync source directory is empty (it is a temporary directory, something like `/private/var/folders/4f/h9t4n3x11qq2538wfyyj_szh0000gn/T/tmp-66668-4Xh5jZhwVXe3/project/.garden/build/simple`)

The module code in question can be found in `garden/core/test/data/test-projects/container/simple-service`.
